app/Main.hs

```diff
@@ -237,17 +239,66 @@ getHttpManager opts = newManager managerSettings
 --
 -- Signals failure through a value of type VaultError, but can also
 -- throw HTTP exceptions.
-vaultEnv :: Context -> ExceptT VaultError IO [EnvVar]
+vaultEnv :: Context -> IO (Either VaultError [EnvVar])
 vaultEnv context = do
-  mountInfo <- requestMountInfo context
-  secrets <- mapExceptT (fmap $ first SecretFileError) $ readSecretList secretFile
-  secretEnv <- requestSecrets context mountInfo secrets
-  checkNoDuplicates (buildEnv secretEnv)
+  mountInfo <- doWithRetries getMountInfo
+  case mountInfo of
+    Left vaultError -> return $ Left vaultError
+    Right mountInfo' -> do
+      secrets <- readSecretsFile secretFile
+      case secrets of
+        Left sfError -> return $ Left $ SecretFileError sfError
+        Right secrets' -> do
+          -- NOTE: This retries fetching all the secrets if fetching one of the
+          -- secrets fails due to an error we should retry on. We would ideally
+          -- keep the secrets that were successfully fetched in memory and only
+          -- retry getting the ones we couldn't get yet, as this makes it less
+          -- likely to get a 'thundering herd' of Vaultenv processes all trying
+          -- and failing to get all of their secrets.
+          -- This would then need to happen in requestSecrets, but I don't want
+          -- to touch that right now.
+          secretEnv <- doWithRetries (getSecrets mountInfo' secrets')
+          case secretEnv of
+            Left vaultError -> return $ Left vaultError
+            Right secretEnv' -> return $ checkNoDuplicates (buildEnv secretEnv')
     where
+      doWithRetries :: (Retry.RetryStatus -> IO (Either VaultError a)) -> IO (Either VaultError a)
+      doWithRetries = Retry.retrying retryPolicy isRetryableFailure
+
+      -- | Indicator function for retrying to retry on VaultErrors (Lefts) that
+      -- shouldRetry thinks we should retry on. Needs to be in IO because the
+      -- actions to perform are in IO as well.
+      isRetryableFailure :: Retry.RetryStatus -> Either VaultError a -> IO Bool
+      isRetryableFailure _retryStatus (Right _) = pure False
+      isRetryableFailure retryStatus (Left err) = shouldRetry retryStatus err
+
+      retryPolicy = vaultRetryPolicy (cCliOptions context)
+
+      getMountInfo _retryStatus = catch (requestMountInfo context) httpErrorHandler
+      getSecrets mountInfo secrets _retryStatus = catch (requestSecrets context mountInfo secrets) httpErrorHandler
```

Comment:
Please add type signatures for both of these.

Reply:
Done.